### PR TITLE
Border agent proxy

### PR DESCRIPTION
**Pull request by bukepo**
This is a proposal implementation of border agent proxy based on design of #1545. Instead of using IP address, RLOC16 is used which allows border agent on host side work without know the mesh local prefix.
* [x] border agent API
* [x] border agent proxy stream
* [x] add new capability definition for border agent proxy

### CODE REVIEW

File: include/openthread/border_agent_proxy.h [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/*
+ *  Copyright (c) 2016, The OpenThread Authors.
```
Comment (nibanks) [inline include/openthread/border_agent_proxy.h]: All these should be 2017 now.

File: src/core/thread/thread_netif.hpp [hunk 2]
```diff
@@ -318,6 +319,7 @@ class ThreadNetif: public Ip6::Netif
      */
     Utils::JamDetector &GetJamDetector(void) { return mJamDetector; }
 #endif // OPENTHREAD_ENABLE_JAM_DETECTION
+    MeshCoP::BorderAgentProxy &GetBorderAgentProxy(void) { return mBorderAgentProxy; }
```
Comment (nibanks) [inline src/core/thread/thread_netif.hpp]: All of this should be optional/configurable and disabled (for now) on Windows.

Comment (bukepo) [inline src/core/thread/thread_netif.hpp]: It's optional now.

**Comment by darconeous** [comment]
I think using the RLOC instead of the full IPv6 address is fine.

**Comment by bukepo** [comment]
I found all STREAM properties are grouped together in current implementation, so `PROP_THREAD_BA_PROXY_STREAM` is renamed to `PROP_STREAM_BA_PROXY` as you see in the code. Do you agree the naming for the new properties. @darconeous

**Comment by bukepo** [comment]
Any more comments? @nibanks @abtink

File: src/ncp/ncp_base.cpp [hunk 3]
```diff
@@ -5129,6 +5283,46 @@ ThreadError NcpBase::SetPropertyHandler_MAC_SRC_MATCH_EXTENDED_ADDRESSES(uint8_t
...
+        otOperationalDataset dataset;
+        otDatasetGetActive(mInstance, &dataset);
+        uint64_t activeTimestamp = dataset.mActiveTimestamp;
+        memset(&dataset, 0, sizeof(dataset));
+        memcpy(dataset.mPSKc.m8, ptr, sizeof(spinel_thread_pskc_t));
+        dataset.mIsPSKcSet = true;
+        dataset.mActiveTimestamp = activeTimestamp;
+        dataset.mIsActiveTimestampSet = true;
+
+        SuccessOrExit(errorCode = otDatasetSetActive(mInstance, &dataset));
```
Comment (abtink) [inline src/ncp/ncp_base.cpp]: Not sure about `SuccessOrExit()` here. If the `otDatasetSetActive`  fails, we need to send a spinel response with the error code. 

The `errorCode` we return from `set` handlers in `NcpBase` are actually any error that may happen in sending/forming the spinel response to the set command, and not any error for the actual set operation.

Comment (bukepo) [inline src/ncp/ncp_base.cpp]: OK, since the implementation for setting PSKc should be the same as other network parameter such as PANID. I'm going to submit a separate PR for that. This PR will focus on border agent proxy feature only.

File: src/ncp/ncp_base.cpp [hunk 4]
```diff
@@ -4493,6 +4566,75 @@ ThreadError NcpBase::SetPropertyHandler_STREAM_NET_INSECURE(uint8_t header, spin
...
+    unsigned int frame_len(0);
+    uint16_t rloc;
+    uint16_t port;
+
+    // STREAM_BA_PROXY requires layer 2 security.
+    otMessage *message = otIp6NewMessage(mInstance, true);
+
+    if (message == NULL)
+    {
+        errorCode = kThreadError_NoBufs;
```
Comment (aeliot) [inline src/ncp/ncp_base.cpp]: Should this use one of the error handling macros instead of an if?
```
VerifyOrExit(message, errorCode = kThreadError_NoBufs);
...
...
exit:
return errorCode;
```

**Comment by darconeous** [comment]
> I found all STREAM properties are grouped together in current implementation, so PROP_THREAD_BA_PROXY_STREAM is renamed to PROP_STREAM_BA_PROXY as you see in the code. Do you agree the naming for the new properties. @darconeous

No, that grouping is for technology-agnostic streams. This stream is thread-specific, and thus should go into a network-technology-specific group.

File: src/ncp/ncp_base.cpp (hunk 4 above)
Comment (abtink) [inline src/ncp/ncp_base.cpp]: It's a bit different in the `Ncp` set handlers. If the operation fails, we still need to report the failure status in a spinel response to host, so we should not go to exit and return this error code. The `errorCode` we return from the `set` handler is actually any error in sending the spinel response (not the error in performing the operation). 

`SetPropertyHandler_MAC_SRC_MATCH_EXTENDED_ADDRESSES()` is example code of where we do use two different variables to track the error and use `VerifyOrExit`:
-  `errorStatus` for error of the set operation (to pass in spinel response). 
- `errorCode` for the error to return from the `SetProp` method.

Comment (bukepo) [inline src/ncp/ncp_base.cpp]: Thanks. I also prefer VerifyOrExit() way. The current implementation follows those existing stream properties' implementation to reduce the chance of bugs. @aeliot @abtink

Comment (jwhui) [inline src/ncp/ncp_base.cpp]: I think we should leave this as is given that this matches the convention used across this class.

That said, `VerifyOrExit()` could be used here and sending the Spinel response to host could be done after the associated `exit:` label.

Comment (jwhui) [inline src/ncp/ncp_base.cpp]: @bukepo, I did not mean to ask you to change to `VerifyOrExit()` since the rest of this class does not follow that design pattern.

Comment (bukepo) [inline src/ncp/ncp_base.cpp]: Sorry for the misunderstanding, I've changed back.

**Comment by darconeous** [comment]
@bukepo: You are still working to add the capability, right?

File: configure.ac [hunk 5]
```diff
@@ -536,6 +536,37 @@ AC_MSG_RESULT(${enable_builtin_mbedtls})
...
+            AC_MSG_ERROR([Invalid value ${enable_border_agent_proxy} for --enable-border-agent-proxy])
+            ;;
+        esac
+    ],
+    [enable_border_agent_proxy=no])
+
+if test "$enable_border_agent_proxy" = "yes"; then
+    OPENTHREAD_ENABLE_BORDER_AGENT_PROXY=1
+else
+    OPENTHREAD_ENABLE_BORDER_AGENT_PROXY=0
```
Comment (nibanks) [inline configure.ac]: We should explicitly define this to 0 in the Windows config file.

**Comment by bukepo** [comment]
@darconeous OK, I will add tomorrow.

File: configure.ac (hunk 5 above)
Comment (bukepo) [inline configure.ac]: I explicitly defined this to 0 on Windows config files I found.

**Comment by bukepo** [comment]
@darconeous I have implemented the capability. Could you please review again?

The range for CAP_THREAD_* is now defined from 1024 to 1152. Is that OK?

**Comment by darconeous** [comment]
Looks like there are merge conflicts, though.

**Comment by bukepo** [comment]
@jwhui just rebased to master. Thanks @darconeous .

File: src/core/meshcop/border_agent_proxy.cpp [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        *static_cast<Coap::Header *>(aHeader), *static_cast<Message *>(aMessage),
+        *static_cast<const Ip6::MessageInfo *>(aMessageInfo));
+}
+
+void BorderAgentProxy::HandleResponse(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                      const otMessageInfo *aMessageInfo, ThreadError aResult)
+{
+    if (aResult != kThreadError_None)
+    {
+        return;
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.cpp]: nit: single exit point per method.

File: src/core/meshcop/border_agent_proxy.hpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace MeshCoP {
+
+class BorderAgentProxy
+{
+public:
+    /**
+     * This constructor initializes the BorderAgentProxy object.
+     *
+     * @param[in]  aThreadNetif  A reference to the Thread network interface.
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.hpp]: nit: missing documentation for `aCoapServer` and `aCoapClient`.

File: src/core/meshcop/border_agent_proxy.hpp [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * This method stops the BorderAgentProxy service.
+     *
+     * @retval kThreadError_None  Successfully stopped the BorderAgentProxy service.
+     *
+     */
+    ThreadError Stop(void);
+
+    ThreadError Send(Message &aMessage, uint16_t aLocator, uint16_t aPort);
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.hpp]: nit: missing Doxygen block

File: src/core/meshcop/border_agent_proxy.hpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+    bool IsEnabled(void) const;
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.hpp]: nit: missing Doxygen block

File: src/core/meshcop/border_agent_proxy.hpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+private:
+
+    static void HandleRelayReceive(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                   const otMessageInfo *aMessageInfo);
+
+    static void HandleResponse(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                               const otMessageInfo *aMessageInfo, ThreadError aResult);
+
+    void DelieverMessage(Coap::Header &aHeader, Message &aMessage, const Ip6::MessageInfo &aMessageInfo);
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.hpp]: typo: `DelieverMessage` -> `DeliverMessage`.

File: src/core/meshcop/border_agent_proxy.hpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+     * @param[in]  aThreadNetif  A reference to the Thread network interface.
+     *
+     */
+    BorderAgentProxy(const Ip6::Address &aMeshLocal16, Coap::Server &aCoapServer, Coap::Client &aCoapClient);
+
+    /**
+     * This method starts the BorderAgentProxy service.
+     *
+     * @retval kThreadError_None  Successfully started the BorderAgentProxy service.
+     *
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.hpp]: nit: document other possible return values.

File: src/core/meshcop/border_agent_proxy.hpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+     */
+    ThreadError Start(otBorderAgentProxyStreamHandler aBorderAgentProxyStreamHandler, void *aContext);
+
+    /**
+     * This method stops the BorderAgentProxy service.
+     *
+     * @retval kThreadError_None  Successfully stopped the BorderAgentProxy service.
+     *
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.hpp]: nit: document other possible return values.

File: configure.ac (hunk 5 above)
Comment (bukepo) [inline configure.ac]: @nibanks do you have other comments?

File: src/ncp/ncp_base.cpp (hunk 4 above)
Comment (bukepo) [inline src/ncp/ncp_base.cpp]: @abtink any more comments?

File: include/openthread/border_agent_proxy.h [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+/**
+ * Send packet through border agent proxy.
+ *
+ * @param[in]  aInstance            A pointer to an OpenThread instance.
+ * @param[in]  aMessage             A pointer to the CoAP Message.
+ * @param[in]  aLocator             Rloc of destination.
+ * @param[in]  aPort                Port of destination.
+ *
+ * @retval kThreadErrorNone         Successfully stopped the border agent proxy.
+ *
```
Comment (jwhui) [inline include/openthread/border_agent_proxy.h]: Also need to document other possible errors here.

File: src/core/meshcop/border_agent_proxy.cpp [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+#if OPENTHREAD_FTD && OPENTHREAD_ENABLE_BORDER_AGENT_PROXY
+
+namespace Thread {
+namespace MeshCoP {
+
+BorderAgentProxy::BorderAgentProxy(const Ip6::Address &aMeshLocal16, Coap::Server &aCoapServer,
+                                   Coap::Client &aCoapClient):
+    mRelayReceive(OPENTHREAD_URI_RELAY_RX, &BorderAgentProxy::HandleRelayReceive, this),
+    mBorderAgentProxyStreamHandler(NULL),
```
Comment (jwhui) [inline src/core/meshcop/border_agent_proxy.cpp]: nit: could use a shorter name `mStreamHandler` since this is already the `BorderAgentProxy` class.

File: etc/visual-studio/ot-ncp-uart.vcxproj [hunk 15]
```diff
@@ -43,6 +43,7 @@
         OPENTHREAD_MULTIPLE_INSTANCE;
         OPENTHREAD_ENABLE_NCP_SPI=0;
         OOPENTHREAD_ENABLE_NCP_UART=1;
+        OPENTHREAD_ENABLE_BORDER_AGENT_PROXY=0;
```
Comment (nibanks) [inline etc/visual-studio/ot-ncp-uart.vcxproj]: Could we move this to the config file instead of defining it in each project file?

File: etc/visual-studio/ot-ncp-uart.vcxproj [hunk 16]
```diff
@@ -43,6 +43,7 @@
         OPENTHREAD_MULTIPLE_INSTANCE;
         OPENTHREAD_ENABLE_NCP_SPI=0;
         OOPENTHREAD_ENABLE_NCP_UART=1;
```
Comment (nibanks) [inline etc/visual-studio/ot-ncp-uart.vcxproj]: I just noticed this. There is an extra 'O' here.

File: etc/visual-studio/ot-ncp-uart.vcxproj (hunk 15 above)
Comment (bukepo) [inline etc/visual-studio/ot-ncp-uart.vcxproj]: moved to `include/openthread-windows-config.h`.

File: include/openthread/border_agent_proxy.h (hunk 13 above)
Comment (bukepo) [inline include/openthread/border_agent_proxy.h]: updated the comments.

File: src/core/meshcop/border_agent_proxy.cpp (hunk 14 above)
Comment (bukepo) [inline src/core/meshcop/border_agent_proxy.cpp]: renamed as you suggested, thanks.
